src/main/java/foodwhere/model/review/ReviewContainsKeywordsPredicate.java

```diff
@@ -0,0 +1,62 @@
+package foodwhere.model.review;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import foodwhere.commons.util.StringUtil;
+import foodwhere.model.commons.Name;
+import foodwhere.model.commons.Tag;
+
+/**
+ * Tests that a {@code Review}'s {@code Name} or {@code Tag} matches any of the keywords given for each field.
+ */
+public class ReviewContainsKeywordsPredicate implements Predicate<Review> {
+    private final HashSet<Name> nameKeywords;
+    private final HashSet<Tag> tagKeywords;
+
+    /**
+     * Creates a ReviewContainsKeywordsPredicate where arguments are in List
+     * @param nameKeywords List of name keywords
+     * @param tagKeywords List of tag keywords
+     */
+    public ReviewContainsKeywordsPredicate(List<Name> nameKeywords, List<Tag> tagKeywords) {
+        this.nameKeywords = new HashSet<Name>(nameKeywords);
+        this.tagKeywords = new HashSet<Tag>(tagKeywords);
+    }
+
+    /**
+     * Creates a ReviewContainsKeywordsPredicate where arguments are in Set
+     * @param nameKeywords Set of name keywords
+     * @param tagKeywords Set of tag keywords
+     */
+    public ReviewContainsKeywordsPredicate(Set<Name> nameKeywords, Set<Tag> tagKeywords) {
+        this.nameKeywords = new HashSet<Name>(nameKeywords);
+        this.tagKeywords = new HashSet<Tag>(tagKeywords);
+    }
+
+    @Override
+    public boolean test(Review review) {
+        boolean hasTag = false;
+        for (Tag tag : review.getTags()) {
+            if (tagKeywords.stream().anyMatch(keyword -> StringUtil.containsWordIgnoreCase(tag.tag, keyword.tag))) {
+                hasTag = true;
+                break;
+            }
+        }
+
+        return nameKeywords.stream()
+                .anyMatch(keyword -> StringUtil.containsWordIgnoreCase(review.getName().fullName, keyword.fullName))
+                || hasTag;
+    }
+
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof ReviewContainsKeywordsPredicate // instanceof handles nulls
+                && nameKeywords.equals(((ReviewContainsKeywordsPredicate) other).nameKeywords))
```

Comment:
This equals check seems to be case-sensitive - is that the intention?

Reply:
Compares Names and Tags using equals() in Name and Tag classes, I will update the methods in both classes to make it case-insensitive